File: engine/src/com/magma/engine/stages/GameStage.java
package com.magma.engine.stages;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Input.Keys;
import com.badlogic.gdx.graphics.Camera;
import com.badlogic.gdx.graphics.OrthographicCamera;
import com.badlogic.gdx.graphics.PerspectiveCamera;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.scenes.scene2d.Stage;
import com.magma.engine.collision.CollisionStage;
import com.magma.engine.gfx.camera.CameraBehavior;
import com.magma.engine.gfx.camera.CameraPanBehavior;

// This abstract type of stage is linked to another stage containing the UI
public class GameStage extends CollisionStage {

    protected final Camera camera;
    protected final Stage ui;
    protected final ViewportContext viewports;

    private CameraBehavior cameraBehavior;
    private CameraBehavior cameraBehaviorUI;
    private boolean initialized;

    public GameStage(ViewportContext viewports, SpriteBatch batch) {
        super(viewports.get(), batch);
        this.camera = getCamera();
        this.viewports = viewports;
        this.ui = new Stage(viewports.getUI(), batch);
    }

    public void setCameraBehavior(CameraBehavior behavior) {
        if (cameraBehavior != null) {
            cameraBehavior.remove();
        }
        cameraBehavior = behavior;
        cameraBehavior.addCamera(this);
        addActor(cameraBehavior);
    }

    public void setCameraBehaviorUI(CameraBehavior behavior) {
        if (cameraBehaviorUI != null) {
            cameraBehaviorUI.remove();
        }
        cameraBehaviorUI = behavior;
        cameraBehaviorUI.addCamera(ui);
        ui.addActor(cameraBehaviorUI);
    }

    public OrthographicCamera getOrthoCamera() {
        if (!(camera instanceof OrthographicCamera)) {
            throw new IllegalStateException("Camera is not orthographic!" + camera.getClass().getSimpleName());
        }
        return (OrthographicCamera) camera;
    }

    public PerspectiveCamera getPersCamera() {
        if (!(camera instanceof PerspectiveCamera)) {
            throw new IllegalStateException("Camera is not perspective! It is " + camera.getClass().getSimpleName());
        }
        return (PerspectiveCamera) camera;
    }

    /**
     * Runs one frame after creating the stage, bad practice so avoid!
     */
    public void create() {
    }

    @Override
    public void act(float delta) {
        if (!initialized) {
            initialized = true;
            create();
        }
        if (Gdx.input.isKeyJustPressed(Keys.F7)) {
            setCameraBehaviorUI(new CameraPanBehavior());
        }
        if (Gdx.input.isKeyJustPressed(Keys.F8)) {
            setCameraBehavior(new CameraPanBehavior());
        }
        super.act(delta);
        ui.act(delta);
    }

    @Override
    public void draw() {
        super.draw();
        ui.getViewport().apply();
        ui.draw();
    }

    @Override
    public void dispose() {
        ui.dispose();
        super.dispose();
    }

    public ViewportContext getViewports() {
        return viewports;
    }

    public Stage getUIStage() {
        return ui;
    }
}

File: engine/src/com/magma/engine/stages/GameStage.kt
package com.magma.engine.stages

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.Input
import com.badlogic.gdx.graphics.OrthographicCamera
import com.badlogic.gdx.graphics.PerspectiveCamera
import com.badlogic.gdx.scenes.scene2d.Stage
import com.magma.engine.MagmaGame
import com.magma.engine.collision.CollisionStage
import com.magma.engine.gfx.camera.CameraBehavior
import com.magma.engine.gfx.camera.CameraPanBehavior

// This abstract type of stage is linked to another stage containing the UI
open class GameStage : CollisionStage() {
    val uiStage: Stage = Stage(ViewportContext.ui, MagmaGame.spriteBatch)

    private var cameraBehavior: CameraBehavior? = null
    private var cameraBehaviorUI: CameraBehavior? = null
    private var initialized = false

    fun setCameraBehavior(behavior: CameraBehavior) {
        cameraBehavior?.remove()
        cameraBehavior = behavior
        addActor(behavior)
    }

    fun setCameraBehaviorUI(behavior: CameraBehavior?) {
        if (cameraBehaviorUI != null) {
            cameraBehaviorUI!!.remove()
        }
        cameraBehaviorUI = behavior
        uiStage.addActor(cameraBehaviorUI)
    }

    val orthoCamera: OrthographicCamera
        get() {
            check(camera is OrthographicCamera) { "Camera is not orthographic!" + camera.javaClass.simpleName }
            return camera as OrthographicCamera
        }
    val persCamera: PerspectiveCamera
        get() {
            check(camera is PerspectiveCamera) { "Camera is not perspective! It is " + camera.javaClass.simpleName }
            return camera as PerspectiveCamera
        }

    override fun act(delta: Float) {
        cameraBehavior?.process(orthoCamera,delta)
        cameraBehaviorUI?.process(orthoCamera,delta)
        if (Gdx.input.isKeyJustPressed(Input.Keys.F7)) {
            setCameraBehaviorUI(CameraPanBehavior())
        }
        if (Gdx.input.isKeyJustPressed(Input.Keys.F8)) {
            setCameraBehavior(CameraPanBehavior())
        }
        super.act(delta)
        uiStage.act(delta)
    }

    override fun draw() {
        super.draw()
        uiStage.viewport.apply()
        uiStage.draw()
    }

    override fun dispose() {
        uiStage.dispose()
        super.dispose()
    }
}